hypothesis-python/src/hypothesis/extra/numpy.py

```diff
@@ -392,9 +392,8 @@ def arrays(
     if isinstance(shape, int):
         shape = (shape,)
     shape = tuple(shape)
-    if not shape:
-        if dtype.kind != u"O":
-            return draw(elements)
+    if (not shape) and (dtype.kind != u"O"):
```

Comment:
Two points on this condition:

1. It looks like the "objects must take a slow path" branch here [is a very old decision](https://github.com/HypothesisWorks/hypothesis/commit/f46f74d30af9c02cddceffdf5f6d5219a6a8ac91#diff-b1b7264abd001dcbb650fb47b9206d69R216) (dating to Hypothesis 1.5.0), and the use of `np.full` makes it obselete anyway.
2. I think we could safely extend this to all shapes for which `np.prod(shape) == 1 or np.prod(shape) == 0)`, i.e. those of any shape but only zero or one elements.  That would be a nice performance enhancement for small arrays!

Reply:
Switching from `not shape` to `np.prod(shape) <= 1` created 14 unit test failures.  I am not sure if it is worth figuring out why all these failures happen, if the current approach of `not shape` fixes the bug.

Reply:
Leave it out for now; we can do a performance patch later if we decide its worth it.